feat(notice): add show/hide controller

dancormier opened the pull request:
This PR adds a controller for `.s-notice` components to provide simple show/hide functionality. It takes _heavy_ inspiration from the modal controller.

---

### TODO

- [x] Documentation!
- [x] Tests!
- [x] ~~Consider generalizing this controller to broadly show/hide anything~~
- [x] Provide focus capture functionality for toast
- [x] Add auto dismissal after interval (for toast)
- [x] Add dismissal interval data attribute
- [x] Add option to prevent auto-dismissal
- [x] Implement for `.s-banner`

dancormier committed:
feat(notice): add show/hide controller

dancormier committed:
Add auto-hiding to the controller

dancormier committed:
Make auto-hide opt-in

dancormier committed:
Replace banner demo with controller

For the record, I don't know if this is a good fit. We probably should consolidate these components first, then port the new controller to work in all sorts of contexts.

dancormier committed:
Remove console logs

dancormier committed:
Create a banner-specific controller

dancormier committed:
Turn generalized notice controller into toast controller

dancormier committed:
I'll show you around this alphabet town

dancormier committed:
Add focus capture to toast controller

dancormier committed:
Move example toast

dancormier committed:
Fix banner icon

dancormier committed:
Moving the toast back for diff purposes

dancormier committed:
Register banner controller

dancormier committed:
Add a banner controller test

It doesn't work atm. It will… hopefully.

dancormier committed:
Update s-banner.test.ts

dancormier committed:
cleanup

dancormier committed:
Minor cleanup

balpha commented on lib/ts/controllers/s-toast.ts:
```diff
@@ -0,0 +1,354 @@
+     * Hide the element after a delay
+     */
+    private hideAfterTimeout() {
+        if (this.data.get("prevent-auto-hide") === "true" || this.data.get("hide-after-timeout") === "0") {
```
Sorry for the drive-by review -- I just stumbled over this, but feel free to ignore me.

The data map API is a remnant from Stimulus 1. It's tricky to update existing usages, but for new stuff we should probably use the value API.

Quoting the [v2 release notes](https://github.com/hotwired/stimulus/releases/tag/v2.0.0),

> The data map API from Stimulus 1.0 will continue to work but is no longer documented and should be considered internal.

giamir commented:
@dancormier it took me a whole morning of debugging but I found the issue with the test. 😅
This private member definition in the banner controller file
```ts
 private bannerTarget!: HTMLElement;
```
needs to be changed to a readonly declaration
```ts
 declare readonly bannerTarget: HTMLElement;
```
as per [Stimulus documentation](https://stimulus.hotwired.dev/reference/using-typescript#define-target-properties):
```
The declare keyword avoids overriding the existing Stimulus property, and just defines the type for TypeScript.
```

My (unverified) theory here is that WTR esbuild is converting typescript module on the fly and it (correctly) creates an actual [private member](https://www.typescriptlang.org/docs/handbook/2/classes.html#private) when it compiles. The webpack ts-loader somehow behave slightly differently and this is why controllers like modal, etc... still works in the docs ...and production.

### Side note
I believe there are few things we could simplify from an API perspective. For example why do we ask our consumers to assign a target to the banner element `data-s-banner-target="banner"` when we could simply reach out for the element via `this.element`?
I think there is also some room to extract common functionalities from the modal, toast. I did not look at the code extensively but I can see there are some patterns which are repeated across all files (this is an optimisation though).

This was fun, I learned a thing or two about stimulus today. I am still not in love with it though. 😅
Especially I found the separation of js from the markup quite tedious for testing/debugging compared to other component based architectures.

dancormier committed:
Update docs w/ JS and inline examples

dancormier committed:
Change bannerTarget to a readonly declaration

h/t @giamir https://github.com/StackExchange/Stacks/pull/1216#issuecomment-1381824710

dancormier committed:
Add test for hideBanner()

dancormier committed:
Add visual tests for each banner variant

dancormier committed:
Add baseline images for banner visual tests

dancormier committed:
Update notice docs with toast controller

dancormier committed:
Update hideBanner, showBanner import

dancormier committed:
Change toast declarations to a readonly

dancormier committed:
Add toast unit tests

dancormier committed:
Add toast visual tests; update banner visual tests
